File: src-tauri/src/commands/backup.rs
use super::*;

#[tauri::command]
pub fn get_backup_status(app: tauri::AppHandle) -> std::result::Result<BackupStatus, String> {
    let app_dir = app_data_dir(&app)?;
    backup_service::get_status(&app_dir).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn create_backup_now(
    app: tauri::AppHandle,
    pool: tauri::State<'_, Pool<SqliteConnectionManager>>,
) -> std::result::Result<BackupStatus, String> {
    let app_dir = app_data_dir(&app)?;
    let status =
        backup_service::create_manual_backup(pool.inner(), &app_dir).map_err(|e| e.to_string())?;
    let metadata_json = audit_metadata_json(serde_json::json!({
        "path": status.last_backup_path.as_deref(),
        "syncFolderPath": status.sync_folder_path.as_deref(),
        "googleDriveConnected": status.google_drive_connected,
    }))?;
    record_command_audit(
        pool.inner(),
        "data_export",
        None,
        "backup",
        "Created manual database backup",
        Some(metadata_json),
    )?;
    Ok(status)
}

#[tauri::command]
pub fn list_backups(app: tauri::AppHandle) -> std::result::Result<Vec<BackupSummary>, String> {
    let app_dir = app_data_dir(&app)?;
    backup_service::list_backups(&app_dir).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn choose_backup_sync_folder(
    app: tauri::AppHandle,
) -> std::result::Result<BackupStatus, String> {
    let app_dir = app_data_dir(&app)?;
    let Some(folder_path) = pick_folder(&app)? else {
        return backup_service::get_status(&app_dir).map_err(|e| e.to_string());
    };

    backup_service::set_sync_folder(&app_dir, Some(folder_path)).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn clear_backup_sync_folder(
    app: tauri::AppHandle,
) -> std::result::Result<BackupStatus, String> {
    let app_dir = app_data_dir(&app)?;
    backup_service::set_sync_folder(&app_dir, None).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn connect_google_drive_backup(
    app: tauri::AppHandle,
) -> std::result::Result<BackupStatus, String> {
    let app_dir = app_data_dir(&app)?;
    backup_service::connect_google_drive(&app_dir).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn disconnect_google_drive_backup(
    app: tauri::AppHandle,
) -> std::result::Result<BackupStatus, String> {
    let app_dir = app_data_dir(&app)?;
    backup_service::disconnect_google_drive(&app_dir).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn upload_latest_backup_to_google_drive(
    app: tauri::AppHandle,
) -> std::result::Result<BackupStatus, String> {
    let app_dir = app_data_dir(&app)?;
    backup_service::upload_latest_backup_to_google_drive(&app_dir).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn choose_restore_backup(
    app: tauri::AppHandle,
) -> std::result::Result<Option<BackupPreview>, String> {
    let Some(file_path) = pick_database_file(&app)? else {
        return Ok(None);
    };

    backup_service::preview_backup(&file_path)
        .map(Some)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn restore_backup(
    app: tauri::AppHandle,
    pool: tauri::State<'_, Pool<SqliteConnectionManager>>,
    source_path: String,
) -> std::result::Result<RestoreResult, String> {
    let app_dir = app_data_dir(&app)?;
    let source_path = PathBuf::from(source_path);
    let result = backup_service::restore_backup(pool.inner(), &app_dir, &source_path)
        .map_err(|e| e.to_string())?;
    let metadata_json = audit_metadata_json(serde_json::json!({
        "sourcePath": source_path.to_string_lossy(),
        "preRestoreBackupPath": result.pre_restore_backup_path.as_str(),
        "schemaVersion": result.schema_version,
        "migrated": result.migrated,
    }))?;
    record_command_audit(
        pool.inner(),
        "database",
        None,
        "restore",
        "Restored database backup",
        Some(metadata_json),
    )?;
    Ok(result)
}

File: src/lib/db-rust/backup.ts
import { invoke } from '@tauri-apps/api/core';
import type {
	ExportData,
	BackupSummary,
	BackupStatus,
	BackupPreview,
	RestoreResult,
	AttendanceEvent,
	Student,
	Class
} from '../types';
export type {
	ExportData,
	BackupSummary,
	BackupStatus,
	BackupPreview,
	RestoreResult
} from '../types';

export async function exportAll(): Promise<ExportData> {
	return await invoke('export_all');
}

export async function exportDatabase(): Promise<string> {
	return await invoke('export_database');
}

export async function exportJsonWithFolder(): Promise<string> {
	return await invoke('export_json_with_folder');
}

export async function exportCsvWithFolder(
	events: AttendanceEvent[],
	students: Student[],
	classes: Class[],
	globalLateAfter: string
): Promise<string> {
	return await invoke('export_csv_with_folder', { events, students, classes, globalLateAfter });
}

export async function importAll(payload: ExportData): Promise<void> {
	return await invoke('import_all', { payload });
}

export async function wipeAll(): Promise<void> {
	return await invoke('wipe_all');
}

export async function getBackupStatus(): Promise<BackupStatus> {
	return await invoke('get_backup_status');
}

export async function createBackupNow(): Promise<BackupStatus> {
	return await invoke('create_backup_now');
}

export async function listBackups(): Promise<BackupSummary[]> {
	return await invoke('list_backups');
}

export async function openBackupFolder(): Promise<string> {
	return await invoke('open_backup_folder');
}

export async function chooseBackupSyncFolder(): Promise<BackupStatus> {
	return await invoke('choose_backup_sync_folder');
}

export async function clearBackupSyncFolder(): Promise<BackupStatus> {
	return await invoke('clear_backup_sync_folder');
}

export async function connectGoogleDriveBackup(): Promise<BackupStatus> {
	return await invoke('connect_google_drive_backup');
}

export async function disconnectGoogleDriveBackup(): Promise<BackupStatus> {
	return await invoke('disconnect_google_drive_backup');
}

export async function uploadLatestBackupToGoogleDrive(): Promise<BackupStatus> {
	return await invoke('upload_latest_backup_to_google_drive');
}

export async function chooseRestoreBackup(): Promise<BackupPreview | null> {
	return await invoke('choose_restore_backup');
}

export async function restoreBackup(sourcePath: string): Promise<RestoreResult> {
	return await invoke('restore_backup', { sourcePath });
}
